File: plugins/plugin-browser/typescript/src/services/process-manager.ts
import { type ChildProcess, spawn } from "node:child_process";
import { existsSync } from "node:fs";
import { platform } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { logger } from "@elizaos/core";

export class BrowserProcessManager {
  private process: ChildProcess | null = null;
  private isRunning = false;
  private binaryPath: string | null = null;

  constructor(private serverPort: number = 3456) {
    this.binaryPath = this.findBinary();
  }

  private getBinaryName(): { primary: string; fallback: string } {
    const platformName = platform();
    const arch = process.arch;
    const ext = platformName === "win32" ? ".exe" : "";

    return {
      primary: `browser-server-${platformName}-${arch}${ext}`,
      fallback: `browser-server-${platformName}${ext}`,
    };
  }

  private findBinary(): string | null {
    const moduleDir = dirname(fileURLToPath(import.meta.url));
    const isDocker = process.env.DOCKER_CONTAINER === "true" || existsSync("/.dockerenv");
    const binaryNames = this.getBinaryName();

    const possiblePaths = [
      ...(isDocker
        ? [
            "/usr/local/bin/browser-server",
            "/usr/local/bin/browser-server-linux",
            "/app/browser-server",
            `/app/binaries/${binaryNames.primary}`,
            `/app/binaries/${binaryNames.fallback}`,
          ]
        : []),

      ...(!isDocker ? [join(moduleDir, "../server/dist/index.js")] : []),
      join(moduleDir, "../server/binaries", binaryNames.primary),
      join(moduleDir, "../server/binaries", binaryNames.fallback),
      join(moduleDir, "../../../browser-server", binaryNames.primary),
      join(moduleDir, "../../../browser-server", binaryNames.fallback),
      join(moduleDir, "../../.bin", "browser-server"),
      join(moduleDir, "../server/dist/index.js"),

      ...(isDocker
        ? ["/app/packages/plugin-browser/server/dist/index.js", "/app/browser-server/dist/index.js"]
        : []),
    ];

    for (const path of possiblePaths) {
      if (existsSync(path)) {
        logger.info(`Found browser server at: ${path}`);
        return path;
      }
    }

    const srcPath = join(moduleDir, "../server/src/index.ts");
    if (existsSync(srcPath)) {
      logger.warn("No compiled binary found, will try to run from source with tsx");
      return srcPath;
    }

    logger.error("Could not find browser server binary or source files");
    logger.error(`Searched paths: ${possiblePaths.join(", ")}`);
    return null;
  }

  async start(): Promise<void> {
    if (this.isRunning) {
      logger.warn("Browser server is already running");
      return;
    }

    if (!this.binaryPath) {
      throw new Error("Browser server binary not found - please ensure server is built");
    }

    const binaryPath = this.binaryPath;

    return new Promise((resolve, reject) => {
      const env = {
        ...process.env,
        BROWSER_SERVER_PORT: this.serverPort.toString(),
        NODE_ENV: process.env.NODE_ENV ?? "production",
        BROWSERBASE_API_KEY: process.env.BROWSERBASE_API_KEY,
        BROWSERBASE_PROJECT_ID: process.env.BROWSERBASE_PROJECT_ID,
        OPENAI_API_KEY: process.env.OPENAI_API_KEY,
        ANTHROPIC_API_KEY: process.env.ANTHROPIC_API_KEY,
        BROWSER_HEADLESS: process.env.BROWSER_HEADLESS,
        CAPSOLVER_API_KEY: process.env.CAPSOLVER_API_KEY,
        OLLAMA_BASE_URL: process.env.OLLAMA_BASE_URL ?? "http://ollama:11434",
        OLLAMA_MODEL: process.env.OLLAMA_MODEL ?? "llama3.2-vision",
        DISPLAY: process.env.DISPLAY ?? ":99",
      };

      const isBinary = !binaryPath.endsWith(".js") && !binaryPath.endsWith(".ts");
      const isTypeScript = binaryPath.endsWith(".ts");

      if (isBinary) {
        this.process = spawn(binaryPath, [], { env });
      } else if (isTypeScript) {
        const tsxPath = require.resolve("tsx/cli", { paths: [process.cwd()] });
        this.process = spawn("node", [tsxPath, binaryPath], { env });
      } else {
        this.process = spawn("node", [binaryPath], { env });
      }

      this.process.stdout?.on("data", (data: Buffer) => {
        const message = data.toString().trim();
        logger.debug(`[BrowserServer] ${message}`);

        if (message.includes("listening on port")) {
          this.isRunning = true;
          resolve();
        }
      });

      this.process.stderr?.on("data", (data: Buffer) => {
        logger.error(`[BrowserServer Error] ${data.toString()}`);
      });

      this.process.on("error", (error: Error) => {
        logger.error(`Failed to start browser server: ${error.message}`);
        this.isRunning = false;
        reject(error);
      });

      this.process.on("exit", (code) => {
        logger.info(`Browser server exited with code ${code}`);
        this.isRunning = false;
      });

      this.waitForServer()
        .then(() => resolve())
        .catch((error) => {
          this.isRunning = false;
          if (this.process) {
            this.process.kill("SIGTERM");
          }
          reject(error);
        });
    });
  }

  private async waitForServer(): Promise<void> {
    const maxAttempts = 30;
    const delay = 1000;

    for (let i = 0; i < maxAttempts; i++) {
      try {
        const ws = require("ws");
        const wsConnection = new ws(`ws://localhost:${this.serverPort}`);

        await new Promise<void>((resolve, reject) => {
          const timeout = setTimeout(() => {
            wsConnection.close();
            reject(new Error("Connection timeout"));
          }, 5000);

          wsConnection.on("open", () => {
            clearTimeout(timeout);
            wsConnection.close();
            resolve();
          });

          wsConnection.on("error", (error: Error) => {
            clearTimeout(timeout);
            reject(error);
          });
        });

        logger.info("Browser server is ready");
        return;
      } catch {}

      await new Promise((resolve) => setTimeout(resolve, delay));
    }

    throw new Error("Browser server failed to start");
  }

  async stop(): Promise<void> {
    if (!this.process || !this.isRunning) {
      return;
    }

    return new Promise((resolve) => {
      this.process?.on("exit", () => {
        this.isRunning = false;
        resolve();
      });

      this.process?.kill("SIGTERM");

      setTimeout(() => {
        if (this.isRunning && this.process) {
          this.process.kill("SIGKILL");
        }
      }, 5000);
    });
  }

  isServerRunning(): boolean {
    return this.isRunning;
  }

  getServerUrl(): string {
    return `ws://localhost:${this.serverPort}`;
  }
}

File: plugins/plugin-browser/python/elizaos_browser/services/process_manager.py
"""Browser server process manager for Python."""

import asyncio
import logging
import os
import subprocess
from pathlib import Path

logger = logging.getLogger(__name__)


class BrowserProcessManager:
    """Manages the browser server process."""

    def __init__(self, server_port: int = 3456) -> None:
        self.server_port = server_port
        self.process: subprocess.Popen[bytes] | None = None
        self.is_running = False
        self.server_path = self._find_server()

    def _find_server(self) -> Path | None:
        """Find the browser server executable or script."""
        # Get the directory of this file
        this_dir = Path(__file__).parent.parent.parent

        # Possible server locations
        possible_paths = [
            # Stagehand server dist
            this_dir.parent.parent / "stagehand-server" / "dist" / "index.js",
            # Relative to plugin-browser
            this_dir.parent / "stagehand-server" / "dist" / "index.js",
            # From workspace root
            Path.cwd() / "plugins" / "plugin-browser" / "stagehand-server" / "dist" / "index.js",
        ]

        for path in possible_paths:
            if path.exists():
                logger.info(f"Found browser server at: {path}")
                return path

        # Check for TypeScript source as fallback
        ts_paths = [
            this_dir.parent.parent / "stagehand-server" / "src" / "index.ts",
            this_dir.parent / "stagehand-server" / "src" / "index.ts",
            Path.cwd() / "plugins" / "plugin-browser" / "stagehand-server" / "src" / "index.ts",
        ]

        for path in ts_paths:
            if path.exists():
                logger.warn(f"Found TypeScript source at: {path} - will need tsx to run")
                return path

        logger.error("Could not find browser server")
        logger.error(f"Searched paths: {possible_paths + ts_paths}")
        return None

    async def start(self) -> None:
        """Start the browser server process."""
        if self.is_running:
            logger.warning("Browser server is already running")
            return

        if not self.server_path:
            raise RuntimeError(
                "Browser server not found - please build stagehand-server:\n"
                "  cd plugins/plugin-browser/stagehand-server && npm install && npm run build"
            )

        env = {
            **os.environ,
            "BROWSER_SERVER_PORT": str(self.server_port),
            "NODE_ENV": os.environ.get("NODE_ENV", "production"),
        }

        # Copy relevant env vars
        for key in [
            "BROWSERBASE_API_KEY",
            "BROWSERBASE_PROJECT_ID",
            "OPENAI_API_KEY",
            "ANTHROPIC_API_KEY",
            "BROWSER_HEADLESS",
            "CAPSOLVER_API_KEY",
            "OLLAMA_BASE_URL",
            "OLLAMA_MODEL",
        ]:
            if key in os.environ:
                env[key] = os.environ[key]

        logger.info(f"Starting browser server from: {self.server_path}")

        if str(self.server_path).endswith(".ts"):
            # TypeScript source - use npx tsx
            cmd = ["npx", "tsx", str(self.server_path)]
        else:
            # JavaScript - use node directly
            cmd = ["node", str(self.server_path)]

        try:
            self.process = subprocess.Popen(
                cmd,
                env=env,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                cwd=self.server_path.parent.parent,  # stagehand-server directory
            )

            # Start output readers
            asyncio.create_task(self._read_output())

            # Wait for server to be ready
            await self._wait_for_server()
            self.is_running = True
            logger.info("Browser server started successfully")

        except Exception as e:
            logger.error(f"Failed to start browser server: {e}")
            self.stop()
            raise

    async def _read_output(self) -> None:
        """Read and log server output."""
        if not self.process:
            return

        while self.process.poll() is None:
            if self.process.stdout:
                line = self.process.stdout.readline()
                if line:
                    logger.debug(f"[BrowserServer] {line.decode().strip()}")
            await asyncio.sleep(0.1)

    async def _wait_for_server(self, max_attempts: int = 30, delay: float = 1.0) -> None:
        """Wait for the server to be ready."""
        import websockets

        url = f"ws://localhost:{self.server_port}"

        for attempt in range(1, max_attempts + 1):
            try:
                async with websockets.connect(url):
                    logger.info("Browser server is ready")
                    return
            except Exception as e:
                if attempt < max_attempts:
                    logger.debug(f"Server not ready yet (attempt {attempt}/{max_attempts}): {e}")
                    await asyncio.sleep(delay)

                    # Check if process died
                    if self.process and self.process.poll() is not None:
                        stderr = self.process.stderr.read().decode() if self.process.stderr else ""
                        raise RuntimeError(f"Browser server process died. stderr: {stderr}")

        raise RuntimeError(f"Browser server failed to start after {max_attempts} attempts")

    def stop(self) -> None:
        """Stop the browser server process."""
        if not self.process:
            return

        logger.info("Stopping browser server")

        try:
            self.process.terminate()
            self.process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            self.process.kill()
            self.process.wait()

        self.process = None
        self.is_running = False

    def __del__(self) -> None:
        """Cleanup on deletion."""
        self.stop()
